File: sdk/python/tempmail_sdk/providers/temporam.py
"""
Temporam 渠道实现
API: https://temporam.com/api/domains, /api/emails
"""

import secrets
import string
from typing import Optional
from urllib.parse import quote

from .. import http as tm_http
from ..normalize import normalize_email
from ..types import Email, EmailInfo

CHANNEL = "temporam"
ORIGIN = "https://temporam.com"

HEADERS = {
    "Accept": "application/json",
    "Accept-Encoding": "gzip, deflate, br",
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
        "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36"
    ),
}


def _random_local(length: int = 18) -> str:
    chars = string.ascii_lowercase + string.digits
    return "sdk" + "".join(secrets.choice(chars) for _ in range(length))


def _get_json(path: str) -> dict:
    resp = tm_http.get(f"{ORIGIN}{path}", headers=HEADERS)
    resp.raise_for_status()
    return resp.json()


def _domains() -> list:
    data = _get_json("/api/domains")
    domains = [
        (item.get("domain") or "").strip()
        for item in data.get("data", [])
        if isinstance(item, dict)
    ]
    domains = [domain for domain in domains if domain]
    if not domains:
        raise RuntimeError("temporam: domain list is empty")
    return domains


def _normalize(raw: dict, email: str) -> Email:
    flat = {
        **raw,
        "id": raw.get("id") or raw.get("uuid") or "",
        "from": raw.get("fromEmail") or raw.get("from") or "",
        "to": raw.get("toEmail") or raw.get("to") or email,
        "date": raw.get("createdAt") or raw.get("created_at") or raw.get("date") or "",
    }
    return normalize_email(flat, email)


def generate_email(domain: Optional[str] = None, **kwargs) -> EmailInfo:
    """创建临时邮箱"""
    domains = _domains()
    selected = secrets.choice(domains)
    if domain:
        if domain not in domains:
            raise RuntimeError(f"temporam: unsupported domain {domain}")
        selected = domain
    email = f"{_random_local()}@{selected}"
    return EmailInfo(channel=CHANNEL, email=email, _token=email)


def get_emails(token: str = "", email: str = "", **kwargs) -> list:
    """获取邮件列表"""
    recipient = token or email
    if not recipient:
        raise RuntimeError("temporam: missing email")
    data = _get_json(f"/api/emails?email={quote(recipient)}&limit=50")
    rows = data.get("data") or []
    out = []
    for row in rows:
        if not isinstance(row, dict):
            continue
        raw = row
        msg_id = row.get("id") or row.get("uuid") or ""
        if msg_id:
            try:
                detail = _get_json(f"/api/emails/{quote(str(msg_id), safe='')}")
                if isinstance(detail.get("data"), dict):
                    raw = detail["data"]
            except Exception:
                raw = row
        out.append(_normalize(raw, recipient))
    return out

File: sdk/ruby/lib/tempmail_sdk/providers/temporam.rb
# frozen_string_literal: true

require "securerandom"

module TempmailSdk
  module Providers
    # Temporam 渠道实现
    # API: https://temporam.com/api
    module Temporam
      CHANNEL = "temporam"
      ORIGIN = "https://temporam.com"

      HEADERS = {
        "Accept" => "application/json",
        "User-Agent" => "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36"
      }.freeze

      module_function

      # 生成随机用户名
      def random_local(length = 18)
        chars = ("a".."z").to_a + ("0".."9").to_a
        "sdk" + Array.new(length) { chars.sample }.join
      end

      # GET 请求并解析 JSON
      def get_json(path)
        resp = Http.get("#{ORIGIN}#{path}", headers: HEADERS)
        resp.raise_for_status
        resp.json
      end

      # 获取可用域名列表
      def fetch_domains
        data = get_json("/api/domains")
        domains = (data["data"] || []).select { |item| item.is_a?(Hash) }.map { |item| (item["domain"] || "").strip }.reject(&:empty?)
        raise "temporam: 域名列表为空" if domains.empty?

        domains
      end

      # 将原始邮件映射为标准化格式
      def normalize_raw(raw, email)
        flat = raw.merge(
          "id" => raw["id"] || raw["uuid"] || "",
          "from" => raw["fromEmail"] || raw["from"] || "",
          "to" => raw["toEmail"] || raw["to"] || email,
          "date" => raw["createdAt"] || raw["created_at"] || raw["date"] || ""
        )
        Normalize.normalize_email(flat, email)
      end

      # 创建临时邮箱
      # @param domain [String, nil]
      # @return [EmailInfo]
      def generate_email(domain = nil)
        domains = fetch_domains
        if domain && !domain.empty?
          raise "temporam: 不支持的域名 #{domain}" unless domains.include?(domain)

          selected = domain
        else
          selected = domains.sample
        end
        email = "#{random_local}@#{selected}"
        EmailInfo.new(channel: CHANNEL, email: email, token: email)
      end

      # 获取邮件列表
      # @param token [String]
      # @param email [String]
      # @return [Array<Email>]
      def get_emails(token, email)
        recipient = token.to_s.empty? ? email : token
        raise "temporam: 缺少邮箱地址" if recipient.to_s.strip.empty?

        data = get_json("/api/emails?email=#{URI.encode_www_form_component(recipient)}&limit=50")
        rows = data["data"] || []
        rows.select { |row| row.is_a?(Hash) }.map do |row|
          raw = row
          msg_id = (row["id"] || row["uuid"] || "").to_s
          unless msg_id.empty?
            begin
              detail = get_json("/api/emails/#{URI.encode_www_form_component(msg_id)}")
              raw = detail["data"] if detail["data"].is_a?(Hash)
            rescue StandardError
              # 详情获取失败时使用列表数据
            end
          end
          normalize_raw(raw, recipient)
        end
      end
    end
  end
end
